chain-tx-enclave/tx-query-next/enclave-app/src/handler/decryption_request.rs

```diff
@@ -0,0 +1,122 @@
+        .map_err(|err| format!("Error while writing request to ZeroMQ: {}", err))?;
+
+    let result_len = zmq_stream
+        .read(&mut result_buf)
+        .map_err(|err| format!("Error while reading response from ZeroMQ: {}", err))?;
+
+    if result_len > max_result_len {
+        return Err("Not enough allocated space to return the sealed tx data".to_owned());
+    }
+
+    match EnclaveResponse::decode(&mut result_buf.as_slice()) {
+        Ok(EnclaveResponse::GetSealedTxData(Some(sealed_logs))) => {
+            let txids = decryption_request.body.txs.clone();
+            let view_key = decryption_request.body.view_key;
+            let mut return_result = Vec::with_capacity(sealed_logs.len());
+
+            for (txid, sealed_log) in txids.into_iter().zip(sealed_logs.into_iter()) {
+                let sealed_data = match SealedData::try_copy_from(&sealed_log) {
+                    Some(sealed_data) => sealed_data,
+                    None => {
+                        return Err("Unable to parse sealed data returned from ZeroMQ".to_owned())
+                    }
+                };
+
+                if sealed_data.aes_data.additional_txt != txid {
+                    return Err("Transaction ID does not match in sealed data".to_owned());
+                }
+
+                let mut unsealed_data = sealed_data
+                    .unseal()
+                    .map_err(|e| format!("Error while unsealing sealed data: {:?}", e))?;
+                let otx = TxWithOutputs::decode(&mut unsealed_data.as_slice());
+                let push: bool;
+
+                match &otx {
+                    Ok(TxWithOutputs::Transfer(Tx {
+                        attributes: TxAttributes { allowed_view, .. },
+                        ..
+                    })) => {
+                        push = is_allowed_view(&allowed_view, &view_key);
+                    }
+                    Ok(TxWithOutputs::StakeWithdraw(WithdrawUnbondedTx {
+                        attributes: TxAttributes { allowed_view, .. },
+                        ..
+                    })) => {
+                        push = is_allowed_view(&allowed_view, &view_key);
+                    }
+                    _ => {
+                        return Err("Invalid transaction type".to_owned());
+                    }
+                }
+
+                if push {
+                    return_result.push(otx.unwrap());
+                }
+
+                unsealed_data.zeroize();
+            }
+
+            let decryption_response = DecryptionResponse { txs: return_result };
+            Ok(decryption_response)
+        }
+        Ok(_) => Err("Unexpected response from ZeroMQ".to_owned()),
+        Err(err) => Err(format!(
+            "Error while decoding response from ZeroMQ: {}",
+            err
+        )),
+    }
+}
+
+#[inline]
+fn is_allowed_view(allowed_views: &[TxAccessPolicy], view_key: &PublicKey) -> bool {
+    // TODO: policy != alldata + const eq?
```

Comment:
policies were reduced in the stabilised format, but constant equality will still be necessary to avoid leaking view keys